cpp/src/arrow/compute/row/grouper.cc

```diff
@@ -44,7 +45,406 @@ namespace compute {
+  }
+
+  AnyKeysGroupingSegmenter(std::vector<TypeHolder> key_types, ExecContext* ctx)
+      : key_types_(std::move(key_types)), ctx_(ctx) {}
+
+  template <typename Batch>
+  Result<GroupingSegment> GetNextSegmentImpl(const Batch& batch, int64_t offset) {
+    ARROW_RETURN_NOT_OK(CheckForGetNextSegment(batch, offset, key_types_));
+    // TODO: make Grouper support Reset(), so it can be cached instead of recreated here
+    ARROW_ASSIGN_OR_RAISE(auto grouper, Grouper::Make(key_types_, ctx_));
+    ARROW_ASSIGN_OR_RAISE(auto datum, grouper->Consume(batch, offset));
+    if (datum.is_array()) {
+      const std::shared_ptr<ArrayData>& data = datum.array();
+      ARROW_DCHECK(data->GetNullCount() == 0);
+      ARROW_DCHECK(*data->type == *uint32());  // as long as Grouper::Consume gives uint32
+      const uint32_t* values = data->GetValues<uint32_t>(1);
+      int64_t cursor;
+      for (cursor = 1; cursor < data->length; cursor++) {
+        if (values[0] != values[cursor]) break;
+      }
+      return MakeSegment(batch.length, offset, std::min(cursor, batch.length - offset));
+    } else if (datum.is_chunked_array()) {
+      return GetNextSegmentChunked(datum.chunked_array(), offset);
+    } else {
+      return Status::Invalid("segmenting unsupported datum kind ", datum.kind());
+    }
+  }
+
+  Result<GroupingSegment> GetNextSegment(const ExecSpan& batch, int64_t offset) override {
+    return GetNextSegmentImpl(batch, offset);
+  }
+
+  Result<GroupingSegment> GetNextSegment(const ExecBatch& batch,
+                                         int64_t offset) override {
+    return GetNextSegmentImpl(batch, offset);
+  }
+
+ private:
+  const std::vector<TypeHolder> key_types_;
+  ExecContext* const ctx_;
+};
+
+Status CheckForConsume(const ExecSpan& batch, int64_t& consume_offset,
+                       int64_t& consume_length) {
+  if (consume_offset < 0) {
+    return Status::Invalid("invalid grouper consume offset: ", consume_offset);
+  }
+  if (consume_length < 0) {
+    consume_length = batch.length - consume_offset;
+  }
+  return Status::OK();
+}
+
+}  // namespace
+
+Result<std::unique_ptr<GroupingSegmenter>> GroupingSegmenter::Make(
+    std::vector<TypeHolder> key_types, ExecContext* ctx) {
+  if (key_types.size() == 0) {
+    return NoKeysGroupingSegmenter::Make();
+  } else if (key_types.size() == 1) {
+    const DataType* type = key_types[0].type;
+    if (type != NULLPTR && is_fixed_width(*type)) {
+      return SimpleKeyGroupingSegmenter::Make(std::move(key_types));
+    }
+  }
+  return AnyKeysGroupingSegmenter::Make(std::move(key_types), ctx);
+}
+
+namespace {
+
+struct BaseGrouper : public Grouper {
+  int IndexOfChunk(const ExecBatch& batch) {
+    int i = 0;
+    for (const auto& value : batch.values) {
+      if (value.is_chunked_array()) {
+        return i;
+      }
+      ++i;
+    }
+    return -1;
+  }
+
+  bool HasConsistentChunks(const ExecBatch& batch, int index_of_chunk) {
```

Comment:
The reason for requiring consistent chunks, which is defined as having the same sequence of chunk lengths for each (chunked-array) value in the batch, is that it ensures we can create an `ExecSpan` without fragmenting the chunks. For example, if we had one value with chunks lengths [3, 5] and a second value with chunk lengths [2, 6] then we'd need to fragment into 3 `ExecSpan` of lengths [2, 1, 5] since they each must cover rows from one chunk. Of course, the fragmentation could easily get much worse with many values having inconsistent chunks, and the performance cost of fragmentation would be annoying. So, while the requirement can be lifted using something like `MultipleChunkIterator`, I'm not sure it is worth it. Let me know your thoughts.